proposals/0354-regex-literals.md

```diff
@@ -248,132 +246,123 @@ let regex = Regex {
+A regex literal may be used with a prefix operator, e.g `let r = ^^/x/` is parsed as `let r = ^^(/x/)`. In this case, when encountering operator characters containing `/` in an expression position, the characters up to the first `/` are split into a prefix operator, and regex literal parsing continues as normal.
 
-For example, all of the following will be parsed as regex literals instead of unapplied operators:
+As already discussed, a regex literal may not begin with a space or tab. However this is not sufficient to disambiguate cases such as:
 
 ```swift
-func foo(_ x: (Int, Int) -> Int, _ y: (Int, Int) -> Int) {}
-
-foo(/, /)  // Will become the regex literal `/, /`
-foo(/^, /) // Will become the regex literal `/^, /`
-foo(!/, /) // Will become prefix `!` on the regex literal `/, /`
+// Unapplied '/' in a call to 'reduce':
+let x = array.reduce(1, /) / 5
+let y = array.reduce(1, /) + otherArray.reduce(1, /)
 
-// Also affects cases where the closing '/' is outside the argument list.
-func bar(_ fn: (Int, Int) -> Int, _ x: Int) -> Int { 0 }
-bar(/, 2) + bar(/, 3) // Will become the (invalid) regex literal `/, 2) + bar(/`
+// Prefix '/' with a potential closing '/'.
+foo(/x) / 2
 
-// Ambiguity with right square bracket:
-struct S {
-  subscript(_ fn: (Int, Int) -> Int) -> Int { 0 }
-}
-func baz(_ x: S) -> Int {
-  x[/] + x[/] // Will become the (invalid) regex literal `/] + x[/`
-}
-
-// Ambiguity with an unapplied operator with two `/` characters:
-func baz(_ x: (Int, Int) -> Int) {}
-baz(/^/) // Will become the regex literal `/^/`
+// Unapplied '!/' with a potential closing '/'.
+bar(!/, 1) / 2
 ```
 
-To disambiguate the cases with the `/` operator, you may surround at least the opening `/` with parentheses, e.g:
+In all of these cases, the opening `/` appears in expression position, and there is a potential closing `/`. To avoid source breakage for such cases, one further heuristic is employed. A regex literal will not be parsed if it contains an unbalanced `)`. This takes both escapes and custom character classes into consideration, and therefore only applies to syntax that would already be invalid for a regex. As such, all of the above cases will continue to be parsed as normal.
+
+This additional heuristic also allows for straightforward disambiguation in source breaking cases where the regex is valid. For example, the following cases will become regex literals:
 
 ```swift
-foo((/), /)
-bar((/), 2) + bar(/, 3)
+baz(/, /)  // Will become the regex literal '/, /'
+baz(/^, /) // Will become the regex literal '/^, /'
+baz(!/, /) // Will become prefix '!' on the regex literal '/, /'
 
-func baz(_ x: S) -> Int {
-  x[(/)] + x[/]
-}
-```
+// Will become the regex literal '/] /'
+let d = hasSubscript[/] / 2
```

Comment:
This case does not seem to make sense to me, the opening bracket is not a character operator which means the `[` becomes unbalanced. With this rule, the line would be parsed as keyword `let`, identifier `d`, operator identifier `=`, identifier `hasSubscript`, subscript opening bracket `[`, regular expression `/] /`, literal `2`, which leads to a bunch errors:
- Expected ',' separator
- No exact matches in call to subscript
It would be like writing `let d = hasSubscript[regex 2`

Wouldn't it make more sense to apply the same rule to `]` and `}`  that is applied to parentheses?

Reply:
The issue is that both `/] /` and `/} /` are valid regex patterns. We could ban them, but it's not clear to me that cases such as `hasSubscript[/] / 2` are common enough to warrant that (we haven't encountered any such cases in the source compatibility testing we've done so far). It is however something we could explore if we start encountering cases like this in the wild.